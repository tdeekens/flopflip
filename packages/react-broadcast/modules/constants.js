export const FLAGS_CHANNEL = '@flopflip';
