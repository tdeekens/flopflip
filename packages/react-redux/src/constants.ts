export const STATE_SLICE = '@flopflip';
