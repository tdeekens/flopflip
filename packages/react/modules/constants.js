export const DEFAULT_FLAG_NAME = 'isFeatureEnabled';
export const ALL_FLAGS = '__flopflop_all_flags__';
