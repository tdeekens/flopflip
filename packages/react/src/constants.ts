export const DEFAULT_FLAG_PROP_KEY = 'isFeatureEnabled';
export const DEFAULT_FLAGS_PROP_KEY = 'featureToggles';
export const ALL_FLAGS_PROP_KEY = '@flopflip/flags';
