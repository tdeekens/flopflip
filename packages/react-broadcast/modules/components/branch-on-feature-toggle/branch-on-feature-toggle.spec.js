import React from 'react';
import { mount } from 'enzyme';
import branchOnFeatureToggle from './branch-on-feature-toggle';
import Configure from '../configure';
import memoryAdapter, { updateFlags } from '@flopflip/memory-adapter';
import { FLAGS_CHANNEL } from '../../constants';

const ToggledComponent = () => <div />;
ToggledComponent.displayName = 'ToggledComponent';
const UntoggledComponent = () => <div />;
UntoggledComponent.displayName = 'UntoggledComponent';

const createTestProps = custom => ({
  adapterArgs: {},

  ...custom,
});

describe('without `untoggledComponent', () => {
  const BranchedComponent = branchOnFeatureToggle({ flag: 'flag1' })(
    ToggledComponent
  );
  let props;
  let wrapper;

  let Container;

  describe('when feature is disabled', () => {
    beforeEach(() => {
      props = createTestProps();
      Container = () => <BranchedComponent />;
      wrapper = mount(
        <Configure {...props} adapter={memoryAdapter}>
          <Container />
        </Configure>
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should not render the `ToggledComponent', () => {
      expect(wrapper).not.toRender(ToggledComponent);
    });

    describe('when enabling feature', () => {
      beforeEach(() => {
        updateFlags({ flag1: true });
        wrapper.update();
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the `ToggledComponent', () => {
        expect(wrapper).toRender(ToggledComponent);
      });
    });
  });
});

describe('with `untoggledComponent', () => {
  const BranchedComponent = branchOnFeatureToggle(
    { flag: 'flag1' },
    UntoggledComponent
  )(ToggledComponent);
  let props;
  let wrapper;

  let Container;

  describe('when feature is disabled', () => {
    beforeEach(() => {
      props = createTestProps();
      Container = () => <BranchedComponent />;
      wrapper = mount(
        <Configure {...props} adapter={memoryAdapter}>
          <Container />
        </Configure>
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should not render the `ToggledComponent', () => {
      expect(wrapper).not.toRender(ToggledComponent);
    });

    it('should render the `UntoggledComponent', () => {
      expect(wrapper).toRender(UntoggledComponent);
    });

    describe('when enabling feature', () => {
      beforeEach(() => {
        updateFlags({ flag1: true });
        wrapper.update();
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the `ToggledComponent', () => {
        expect(wrapper).toRender(ToggledComponent);
      });

      it('should not render the `UntoggledComponent', () => {
        expect(wrapper).not.toRender(UntoggledComponent);
      });
    });
  });
});
