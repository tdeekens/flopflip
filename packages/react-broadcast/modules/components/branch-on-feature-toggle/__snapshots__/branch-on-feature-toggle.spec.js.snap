// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with \`untoggledComponent when feature is disabled should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "getIsReady": [Function],
      "reconfigure": [Function],
    }
  }
  adapterArgs={Object {}}
  defaultFlags={Object {}}
  shouldDeferAdapterConfiguration={false}
>
  <ConfigureAdapter
    adapter={
      Object {
        "configure": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
      }
    }
    adapterArgs={
      Object {
        "onFlagsStateChange": [Function],
        "onStatusStateChange": [Function],
      }
    }
    defaultFlags={Object {}}
    render={null}
    shouldDeferAdapterConfiguration={false}
  >
    <branchOnFeatureToggle(ToggledComponent)>
      <withProps(omitProps(branch(ToggledComponent)))
        @flopflip/flags={Object {}}
      >
        <omitProps(branch(ToggledComponent))
          @flopflip/flags={Object {}}
          isFeatureEnabled={false}
        >
          <branch(ToggledComponent)
            isFeatureEnabled={false}
          >
            <renderComponent(UntoggledComponent)
              isFeatureEnabled={false}
            >
              <UntoggledComponent
                isFeatureEnabled={false}
              >
                <div />
              </UntoggledComponent>
            </renderComponent(UntoggledComponent)>
          </branch(ToggledComponent)>
        </omitProps(branch(ToggledComponent))>
      </withProps(omitProps(branch(ToggledComponent)))>
    </branchOnFeatureToggle(ToggledComponent)>
  </ConfigureAdapter>
</ConfigureFlopflip>
`;

exports[`with \`untoggledComponent when feature is disabled when enabling feature should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "getIsReady": [Function],
      "reconfigure": [Function],
    }
  }
  adapterArgs={Object {}}
  defaultFlags={Object {}}
  shouldDeferAdapterConfiguration={false}
>
  <ConfigureAdapter
    adapter={
      Object {
        "configure": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
      }
    }
    adapterArgs={
      Object {
        "onFlagsStateChange": [Function],
        "onStatusStateChange": [Function],
      }
    }
    defaultFlags={Object {}}
    render={null}
    shouldDeferAdapterConfiguration={false}
  >
    <branchOnFeatureToggle(ToggledComponent)>
      <withProps(omitProps(branch(ToggledComponent)))
        @flopflip/flags={
          Object {
            "flag1": true,
          }
        }
      >
        <omitProps(branch(ToggledComponent))
          @flopflip/flags={
            Object {
              "flag1": true,
            }
          }
          isFeatureEnabled={true}
        >
          <branch(ToggledComponent)
            isFeatureEnabled={true}
          >
            <ToggledComponent
              isFeatureEnabled={true}
            >
              <div />
            </ToggledComponent>
          </branch(ToggledComponent)>
        </omitProps(branch(ToggledComponent))>
      </withProps(omitProps(branch(ToggledComponent)))>
    </branchOnFeatureToggle(ToggledComponent)>
  </ConfigureAdapter>
</ConfigureFlopflip>
`;

exports[`without \`untoggledComponent when feature is disabled should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "getIsReady": [Function],
      "reconfigure": [Function],
    }
  }
  adapterArgs={Object {}}
  defaultFlags={Object {}}
  shouldDeferAdapterConfiguration={false}
>
  <ConfigureAdapter
    adapter={
      Object {
        "configure": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
      }
    }
    adapterArgs={
      Object {
        "onFlagsStateChange": [Function],
        "onStatusStateChange": [Function],
      }
    }
    defaultFlags={Object {}}
    render={null}
    shouldDeferAdapterConfiguration={false}
  >
    <branchOnFeatureToggle(ToggledComponent)>
      <withProps(omitProps(branch(ToggledComponent)))
        @flopflip/flags={Object {}}
      >
        <omitProps(branch(ToggledComponent))
          @flopflip/flags={Object {}}
          isFeatureEnabled={false}
        >
          <branch(ToggledComponent)
            isFeatureEnabled={false}
          >
            <Nothing
              isFeatureEnabled={false}
            />
          </branch(ToggledComponent)>
        </omitProps(branch(ToggledComponent))>
      </withProps(omitProps(branch(ToggledComponent)))>
    </branchOnFeatureToggle(ToggledComponent)>
  </ConfigureAdapter>
</ConfigureFlopflip>
`;

exports[`without \`untoggledComponent when feature is disabled when enabling feature should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "getIsReady": [Function],
      "reconfigure": [Function],
    }
  }
  adapterArgs={Object {}}
  defaultFlags={Object {}}
  shouldDeferAdapterConfiguration={false}
>
  <ConfigureAdapter
    adapter={
      Object {
        "configure": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
      }
    }
    adapterArgs={
      Object {
        "onFlagsStateChange": [Function],
        "onStatusStateChange": [Function],
      }
    }
    defaultFlags={Object {}}
    render={null}
    shouldDeferAdapterConfiguration={false}
  >
    <branchOnFeatureToggle(ToggledComponent)>
      <withProps(omitProps(branch(ToggledComponent)))
        @flopflip/flags={
          Object {
            "flag1": true,
          }
        }
      >
        <omitProps(branch(ToggledComponent))
          @flopflip/flags={
            Object {
              "flag1": true,
            }
          }
          isFeatureEnabled={true}
        >
          <branch(ToggledComponent)
            isFeatureEnabled={true}
          >
            <ToggledComponent
              isFeatureEnabled={true}
            >
              <div />
            </ToggledComponent>
          </branch(ToggledComponent)>
        </omitProps(branch(ToggledComponent))>
      </withProps(omitProps(branch(ToggledComponent)))>
    </branchOnFeatureToggle(ToggledComponent)>
  </ConfigureAdapter>
</ConfigureFlopflip>
`;
